### PR TITLE
Feat/#442 implement error insufficient balance in callop

### LINKED ISSUE #442: Implement `ErrorInsufficientBalance` in callop
Check circuit implementation for more details, https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/998

### PR DESCRIPTION
**Pull request by KimiWu123**
fixed #442

### CODE REVIEW

File: src/zkevm_specs/evm_circuit/execution/callop.py [hunk 1]
```diff
@@ -34,12 +32,10 @@ def callop(instruction: Instruction):
         if is_delegatecall == 1
         else (Word(0), Word(0))
     )
-    parent_caller_address = instruction.word_to_address(parent_caller_address_word)
+    instruction.word_to_address(parent_caller_address_word)
```
Comment (rrtoledo) [inline src/zkevm_specs/evm_circuit/execution/callop.py:35]: I'm not sure to understand what you want to do here. Is that an oversight?

**Review by rrtoledo — COMMENTED** [review]
Except for one word_to_address LGTM!

File: src/zkevm_specs/evm_circuit/execution/callop.py (hunk 1 above)
Comment (KimiWu123) [reply src/zkevm_specs/evm_circuit/execution/callop.py]: Nice catch! Fixed in [dec73b2](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/465/commits/dec73b295494a00caf23fba24e42bbad14880fbc)

**Review by rrtoledo — APPROVED** [review]
LGTM! :)

File: tests/evm/test_callop.py [hunk 2]
```diff
@@ -129,15 +129,12 @@ def gen_testing_data():
     ]
     stacks = [
         Stack(),
-        Stack(value=int(1e18)),
-        Stack(gas=100),
-        Stack(gas=100000),
-        Stack(cd_offset=64, cd_length=320, rd_offset=0, rd_length=32),
-        Stack(cd_offset=0, cd_length=32, rd_offset=64, rd_length=320),
```
Comment (ChihChengLiang) [inline tests/evm/test_callop.py:136]: Can you expand on why we are removing these tests?

File: tests/evm/test_callop.py [hunk 3]
```diff
@@ -257,7 +257,9 @@ def test_callop(
         is_empty_code_hash = True
     callee_bytecode_hash = Word(callee_bytecode_hash if not callee.is_empty() else 0)
 
-    is_success = False if callee is CALLEE_WITH_REVERT_BYTECODE else True
+    # Only check balance and stack depth
+    is_precheck_ok = caller.balance >= value and depth < 1025
+    is_success = False if is_precheck_ok is False or callee is CALLEE_WITH_REVERT_BYTECODE else True
```
Comment (ChihChengLiang) [inline tests/evm/test_callop.py:262]: The logic can be simplified.

```suggestion
    is_success = (is_precheck_ok and callee != CALLEE_WITH_REVERT_BYTECODE)
```

**Review by ChihChengLiang — APPROVED** [review]
LGTM after comments are addressed.

File: src/zkevm_specs/evm_circuit/util/call_gadget.py [hunk 4]
```diff
@@ -61,17 +65,23 @@ def __init__(
...
 
-        if self.IS_SUCCESS_CALL == FQ(1):
-            # Verify is_success is a bool
-            instruction.constrain_bool(self.is_success)
-            self.gas = instruction.word_to_fq(gas, N_BYTES_GAS)
-            self.is_u64_gas = instruction.is_zero(instruction.sum(gas.to_le_bytes()[N_BYTES_GAS:]))
-        else:
+        # Verify is_success is a bool
+        instruction.constrain_bool(self.is_success)
+        if self.IS_SUCCESS_CALL == FQ(0):
```
[the remark this answers is not in the document]
Comment (KimiWu123) [reply src/zkevm_specs/evm_circuit/util/call_gadget.py:70]: This is a bad naming and I haven't found a better name. `IS_SUCCESS_CALL` is to distinguish `callop` and `error_oog_call`. In our circuit it delcared like this 
```
pub(crate) struct CommonCallGadget<F, const IS_SUCCESS_CALL: bool>
```
That's also the reason why `IS_SUCCESS_CALL` is all capital.

File: tests/evm/test_callop.py (hunk 2 above)
Comment (KimiWu123) [reply tests/evm/test_callop.py:136]: There are 768 test cases in `test_callop` now and that's why I merged these 5 different stacks into 2 to reduce the number of cases. There will be 1,344 cases if I don't merge them. But the most important is I think it still covers the testing scenarios or scope as it was.

**Pull request merged**